adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/CertificateDeviceResolver.java

```diff
@@ -0,0 +1,302 @@
+        this.authProvider = Objects.requireNonNull(authProvider);
+    }
+
+    private Span newSpan(final String operation) {
+        return tracer.buildSpan(operation)
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_SERVER)
+                .withTag(Tags.COMPONENT.getKey(), adapterName)
+                .start();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public AdditionalInfo getInfo(final Principal clientIdentity, final Object customArgument) {
+
+        if (clientIdentity instanceof X509CertPath) {
+            final Span span = newSpan("X500-get-device-from-certificate");
+            final CompletableFuture<AdditionalInfo> authResult = new CompletableFuture<>();
+            context.runOnContext(go -> {
+                final X509CertPath certPath = (X509CertPath) clientIdentity;
+                validateCertificateAndLoadDevice(certPath.getPath(), span)
+                        .onSuccess(info -> authResult.complete(info))
+                        .onFailure(t -> authResult.completeExceptionally(t));
+            });
+            try {
+                // client will only wait a limited period of time,
+                // so no need to use get(Long, TimeUnit) here
+                final AdditionalInfo info = authResult.join();
+                span.log("resolved authenticated principal");
+                return info;
+            } catch (final CompletionException e) {
+                TracingHelper.logError(span, "could not resolve authenticated principal", e);
+                LOG.debug("could not resolve x509 certificate device [{}]", clientIdentity.getName(), e);
+            } finally {
+                span.finish();
+            }
+        } else {
+            LOG.info("unsupported Principal type: {}", clientIdentity.getClass());
+        }
+        return null;
+    }
+
+    /**
+     * Validate certificate used by a device in a x509 based DTLS handshake and load device.
+     *
+     * @param certChain certificate chain.
+     * @param span tracing span.
+     */
+    private Future<AdditionalInfo> validateCertificateAndLoadDevice(final CertPath certPath, final Span span) {
+        final List<? extends Certificate> list = certPath.getCertificates();
+        final Certificate[] certChain = list.toArray(new Certificate[list.size()]);
+        final Promise<AdditionalInfo> authResult = Promise.promise();
+
+        auth.validateClientCertificate(certChain, span.context())
+                .onSuccess(ar -> {
+                    final SubjectDnCredentials credentials = authProvider.getCredentials(ar);
+                    if (credentials == null) {
+                        authResult.fail("x509: no valid subject-dn!");
+                    } else {
+                        LOG.debug("x509: subject-dn: {}", credentials.getAuthId());
+                        applyTraceSamplingPriority(credentials, span)
+                                .onSuccess(v -> {
```

Comment:
What if *applyTraceSamplingPriority* fails? It looks like *authResult* will not be completed in that case, or am I mistaken?

Reply:
AFAIK, it maps all failures to succeeded.